File: agents/weather_agent.py
"""
WeatherAgent — current weather via wttr.in (no API key required).

Usage:
  @weather-agent [location]   fetch weather for location
  @weather-agent help         show usage

Environment:
  WEATHER_DEFAULT_LOCATION   default location (fallback: London)
"""

from __future__ import annotations

import logging
import os
import time
import urllib.parse

from ..config import CONFIG

from ..core.actor import Actor, Message, MessageType

logger = logging.getLogger(__name__)

_DEFAULT_LOCATION = CONFIG.weather_default_location
_USER_AGENT       = "AgentFlow-WeatherAgent/1.0"
_TIMEOUT_SEC      = 10


def _url_encode(location: str) -> str:
    safe = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.,'"
    out = []
    for ch in location:
        if ch == " ":
            out.append("+")
        elif ch in safe:
            out.append(ch)
        else:
            out.append(urllib.parse.quote(ch, safe=""))
    return "".join(out)


class WeatherAgent(Actor):
    """Real-time weather actor using wttr.in."""

    def __init__(self, **kwargs):
        kwargs.setdefault("name", "weather-agent")
        super().__init__(**kwargs)
        self.protected = False
        self._default_location = _DEFAULT_LOCATION

    async def on_start(self):
        await self._mqtt_publish(
            f"agents/{self.actor_id}/spawn",
            {
                "agentId":   self.actor_id,
                "agentName": self.name,
                "agentType": "data",
                "timestamp": time.time(),
            },
        )
        logger.info(f"[{self.name}] started — default location: {self._default_location}")

    async def handle_message(self, msg: Message):
        if msg.type not in (MessageType.TASK, MessageType.RESULT):
            return

        payload = msg.payload or {}
        if isinstance(payload, dict):
            text = str(payload.get("text") or payload.get("content") or "")
        else:
            text = str(payload)

        for prefix in ("@weather-agent", "@weather_agent"):
            if text.lower().startswith(prefix):
                text = text[len(prefix):].lstrip()
                break

        if text.lower() == "help":
            default = self._default_location
            await self._reply(
                f"**WeatherAgent** — current conditions via wttr.in (no API key needed)\n\n"
                f"```\n"
                f"@weather-agent              # {default} (default)\n"
                f"@weather-agent Tokyo\n"
                f"@weather-agent New York\n"
                f"@weather-agent 48.8566,2.3522  # coordinates\n"
                f"```\n"
                f"Set `WEATHER_DEFAULT_LOCATION` in `.env` to change the default."
            )
            return

        location = text.strip() or self._default_location
        await self._reply(f"🌦 Fetching weather for **{location}**...")
        result = await self._fetch(location)
        await self._reply(result)
        self.metrics.tasks_completed += 1

    async def _fetch(self, location: str) -> str:
        try:
            import aiohttp
        except ImportError:
            return "Error: `aiohttp` is not installed. Cannot fetch weather."

        encoded = _url_encode(location)
        url     = f"https://wttr.in/{encoded}?format=j1"
        timeout = aiohttp.ClientTimeout(total=_TIMEOUT_SEC)
        headers = {"User-Agent": _USER_AGENT}

        try:
            async with aiohttp.ClientSession(timeout=timeout, headers=headers) as session:
                async with session.get(url) as resp:
                    if resp.status != 200:
                        # Fallback to one-line format
                        url2 = f"https://wttr.in/{encoded}?format=3"
                        async with session.get(url2) as r2:
                            return (await r2.text()).strip()
                    data = await resp.json(content_type=None)
                    return self._format(data, location, encoded)
        except Exception as exc:
            logger.warning(f"[{self.name}] weather fetch error: {exc}")
            return f"⚠ Could not fetch weather for '{location}': {exc}"

    @staticmethod
    def _format(data: dict, location: str, encoded: str) -> str:
        try:
            cc   = data["current_condition"][0]
            desc = cc.get("weatherDesc", [{}])[0].get("value", "N/A")
            area = (
                data.get("nearest_area", [{}])[0]
                    .get("areaName", [{}])[0]
                    .get("value", location)
            )
            return (
                f"**Weather in {area}**\n\n"
                f"🌡 **{cc.get('temp_C', '?')}°C / {cc.get('temp_F', '?')}°F**"
                f" (feels like {cc.get('FeelsLikeC', '?')}°C)\n"
                f"☁ {desc}\n"
                f"💧 Humidity: {cc.get('humidity', '?')}%\n"
                f"💨 Wind: {cc.get('windspeedKmph', '?')} km/h {cc.get('winddir16Point', '?')}\n"
                f"👁 Visibility: {cc.get('visibility', '?')} km\n"
                f"☀ UV index: {cc.get('uvIndex', '?')}\n\n"
                f"*Data: [wttr.in](https://wttr.in/{encoded})*"
            )
        except (KeyError, IndexError):
            return f"Received data for **{location}** but could not parse it."

    async def _reply(self, content: str):
        await self._mqtt_publish(
            f"agents/{self.actor_id}/chat",
            {"from": self.name, "to": "user", "content": content, "timestamp": time.time()},
        )

File: node/src/agents/weather_agent.ts
/**
 * WeatherAgent — current weather via wttr.in (Node.js port).
 * No API key required. agentType: "data"
 */

import axios from "axios";
import { Actor, MqttPublisher } from "../core/actor";
import { Message, MessageType } from "../core/types";

const DEFAULT_LOCATION = process.env["WEATHER_DEFAULT_LOCATION"] ?? "London";

function urlEncode(location: string): string {
  return location
    .split("")
    .map((ch) => (ch === " " ? "+" : encodeURIComponent(ch)))
    .join("");
}

export class WeatherAgent extends Actor {
  private _defaultLocation: string;

  constructor(publish: MqttPublisher, actorId?: string) {
    super("weather-agent", publish, actorId);
    this._defaultLocation = DEFAULT_LOCATION;
  }

  protected override async onStart(): Promise<void> {
    this.publishSpawn("data");
  }

  override async handleMessage(msg: Message): Promise<void> {
    if (msg.type !== MessageType.Task && msg.type !== MessageType.Text) return;
    let text = Actor.extractText(msg.payload).trim();
    text = Actor.stripPrefix(text, "@weather-agent", "@weather_agent");

    if (text.toLowerCase() === "help") {
      this.replyChat(
        `**WeatherAgent** — current conditions via wttr.in (no API key needed)\n\n` +
        "```\n" +
        `@weather-agent              # ${this._defaultLocation} (default)\n` +
        "@weather-agent Tokyo\n" +
        "@weather-agent New York\n" +
        "```\n" +
        "Set `WEATHER_DEFAULT_LOCATION` in env to change the default."
      );
      return;
    }

    const location = text.trim() || this._defaultLocation;
    this.replyChat(`🌦 Fetching weather for **${location}**...`);
    const result = await this._fetch(location);
    this.replyChat(result);
    this.metrics.tasksCompleted++;
  }

  private async _fetch(location: string): Promise<string> {
    const encoded = urlEncode(location);
    const url = `https://wttr.in/${encoded}?format=j1`;
    try {
      const resp = await axios.get(url, {
        timeout: 10_000,
        headers: { "User-Agent": "Wactorz-WeatherAgent/1.0" },
        validateStatus: (s) => s < 500,
      });
      if (resp.status !== 200) {
        // fallback: one-line format
        const r2 = await axios.get(`https://wttr.in/${encoded}?format=3`, {
          timeout: 10_000,
          headers: { "User-Agent": "Wactorz-WeatherAgent/1.0" },
        });
        return String(r2.data).trim();
      }
      return this._format(resp.data as Record<string, unknown>, location, encoded);
    } catch (e) {
      return `⚠ Could not fetch weather for '${location}': ${e}`;
    }
  }

  private _format(data: Record<string, unknown>, location: string, encoded: string): string {
    try {
      const cc = (data["current_condition"] as Record<string, unknown>[])[0];
      const desc = ((cc["weatherDesc"] as { value: string }[])[0]?.value) ?? "N/A";
      const area =
        ((data["nearest_area"] as Record<string, unknown>[])[0]?.["areaName"] as { value: string }[])[0]?.value ??
        location;

      return (
        `**Weather in ${area}**\n\n` +
        `🌡 **${cc["temp_C"]}°C / ${cc["temp_F"]}°F** (feels like ${cc["FeelsLikeC"]}°C)\n` +
        `☁ ${desc}\n` +
        `💧 Humidity: ${cc["humidity"]}%\n` +
        `💨 Wind: ${cc["windspeedKmph"]} km/h ${cc["winddir16Point"]}\n` +
        `👁 Visibility: ${cc["visibility"]} km\n` +
        `☀ UV index: ${cc["uvIndex"]}\n\n` +
        `*Data: [wttr.in](https://wttr.in/${encoded})*`
      );
    } catch {
      return `Received data for **${location}** but could not parse it.`;
    }
  }
}
